veps/sig-storage/dic-on-heterogeneous-cluster/dic-on-heterogeneous-cluster.md

```diff
@@ -0,0 +1,459 @@
+and maintain a list of supported architectures. That means that HCO will be responsible for the second aspect: the
+workload node architectures.
+
+The DataImportCron controller in CDI, is the component that monitors the images and that makes it to be responsible for
+the third aspect: the architectures supported by the image.
+
+### Maintaining a List of Supported Architectures
+
+The HCO will be enhanced to track the architectures of workload nodes in the cluster. It will maintain a distinct list
+of these architectures, derived from the `status.nodeInfo.architecture` field of each node.
+
+This list will be dynamically updated as nodes are added or removed from the cluster.
+
+By default, the HCO will identify workload nodes as those labeled with `node-role.kubernetes.io/worker`. If the
+`spec.workloads.nodePlacement` field in the `HyperConverged` CR is defined, HCO will use this field to determine the
+workload nodes instead.
+
+### Reconciling the `SSP` Custom Resource
+
+When reconciling the `SSP` CR, HCO will:
+
+    1. Populate the `spec.cluster.workloadArchitectures` filed in the `SSP` CR with the supported architectures.
+2. Add the `ssp.kubevirt.io/dict.architectures` annotation to each **predefined** `DataImportCronTemplate` object in the
+   `SSP` CR's `spec.commonTemplates.dataImportCronTemplates` field. This annotation will include a comma-separated list
+   of the supported architectures.
+   > **Note**: For the **custom** `DataImportCronTemplate` objects, HCO will not add this annotation. Users are
+   > responsible for adding the annotation to their custom `DataImportCronTemplate` objects, based on the architectures
+   > supported by the specific image.
+
+For example:
+
+```yaml
+apiVersion: ssp.kubevirt.io/v1beta3
+kind: SSP
+spec:
+  cluster:
+    workloadArchitectures:
+      - amd64
+      - arm64
+      - s390x
+  commonTemplates:
+    dataImportCronTemplates:
+      - metadata:
+          name: "centos-stream9-image-cron"
+          annotations:
+            ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"
+    ...
+```
+
+### Reconciling the `DataImportCron` Custom Resources
+
+Upon reconciling the SSP CR, for each `DataImportCronTemplate` object, the SSP operator shall do the following:
+
+1. determine the required architectures. The required architectures are the intersection of
+   the values listed in the `spec.cluster.workloadArchitectures` field, and the
+   `ssp.kubevirt.io/dict.architectures` annotation. For example, if the
+   `ssp.kubevirt.io/dict.architectures` annotation is `"arm64,amd64"` and the
+   `spec.cluster.workloadArchitectures` is `["arm64", "s390x"]`, the required architectures are `["arm64"]`.
+
+   > **Note**: for pre-defined `DataImportCronTemplate` objects, the `ssp.kubevirt.io/dict.architectures`
+   annotation is expected to be the same as the architectures listed in the `spec.cluster.workloadArchitectures` field.
```

Comment:
IMO a DICT should have all the arches it supports in it's annotation, regardless of the architectures available in the cluster right now.

Reply:
HCO does not need to know it, but someone supplies this information to HCO (the DICTs) and they should know.

Reply:
The DICT are based on hard coded files in the HCO image. DICT objects in these files are referring to images in a registry (or in an imageStream). These images are frequently changing - this is why we have the DIC mechanism. 

So I can assume that all the images are actually manifests, that points to a fixed list of arch specific images, and only the specific images are changed. But who can promise that this assumption is valid? The DICT file created at image build time. The image manifest is changing all the time. In theory, it can become a regular image, or maybe the arch list in the manifest will be changed. This is not something HCO can monitor.